File: apps/web/src/lib/transcode.ts
import type { ContainerFormat } from "@/lib/render-settings";

function normalizeFolder(folder: string | null | undefined): string | null {
  if (folder == null) {
    return null;
  }
  const normalized = folder.trim().replace(/^\/+|\/+$/g, "");
  return normalized || null;
}

export function buildTranscodeOutputKey(
  sourceKey: string,
  {
    container,
    prependFolder,
    appendFolder,
  }: {
    container: ContainerFormat | string;
    prependFolder?: string | null;
    appendFolder?: string | null;
  },
): string {
  const normalizedKey = sourceKey.trim().replace(/^\/+|\/+$/g, "");
  if (!normalizedKey) {
    throw new Error("source_key must not be empty.");
  }

  const lastSlash = normalizedKey.lastIndexOf("/");
  const directory = lastSlash >= 0 ? normalizedKey.slice(0, lastSlash) : "";
  const filename = lastSlash >= 0 ? normalizedKey.slice(lastSlash + 1) : normalizedKey;
  const dot = filename.lastIndexOf(".");
  const stem = dot > 0 ? filename.slice(0, dot) : filename;
  const suffix = container.toString().trim().toLowerCase().replace(/^\./, "");
  if (!suffix) {
    throw new Error("container must not be empty.");
  }

  const parts: string[] = [];
  const prepend = normalizeFolder(prependFolder);
  const append = normalizeFolder(appendFolder);
  if (prepend) {
    parts.push(prepend);
  }
  if (directory) {
    parts.push(directory);
  }
  if (append) {
    parts.push(append);
  }
  parts.push(stem);
  return `${parts.join("/")}.${suffix}`;
}

export function formatBytes(size: number): string {
  if (size < 1024) {
    return `${size} B`;
  }
  if (size < 1024 * 1024) {
    return `${(size / 1024).toFixed(1)} KB`;
  }
  if (size < 1024 * 1024 * 1024) {
    return `${(size / (1024 * 1024)).toFixed(1)} MB`;
  }
  return `${(size / (1024 * 1024 * 1024)).toFixed(1)} GB`;
}

export function formatModified(value: string | null): string {
  if (!value) {
    return "—";
  }
  const date = new Date(value);
  if (Number.isNaN(date.getTime())) {
    return value;
  }
  return date.toLocaleString();
}

File: packages/baserender/src/baserender/transcode.py
"""Helpers for direct file transcoding (no OTIO)."""

from __future__ import annotations

from pathlib import PurePosixPath


def _normalize_folder(folder: str | None) -> str | None:
    if folder is None:
        return None
    normalized = folder.strip().strip("/")
    return normalized or None


def build_transcode_output_key(
    source_key: str,
    *,
    container: str,
    prepend_folder: str | None = None,
    append_folder: str | None = None,
) -> str:
    """Build the S3 output key for a transcode job.

    Preserves directory structure under the source key. Optional prepend/append
    folders are inserted before the full source path or before the filename,
    respectively.
    """
    normalized_key = source_key.strip().strip("/")
    if not normalized_key:
        raise ValueError("source_key must not be empty.")

    path = PurePosixPath(normalized_key)
    directory = path.parent.as_posix() if path.parent != PurePosixPath(".") else ""
    stem = path.stem
    suffix = container.strip().lower().lstrip(".")
    if not suffix:
        raise ValueError("container must not be empty.")

    parts: list[str] = []
    prepend = _normalize_folder(prepend_folder)
    append = _normalize_folder(append_folder)
    if prepend:
        parts.append(prepend)
    if directory:
        parts.append(directory)
    if append:
        parts.append(append)
    parts.append(stem)
    return "/".join(parts) + f".{suffix}"
